lib/charms/prometheus_k8s/v0/prometheus_scrape.py

```diff
@@ -545,84 +546,203 @@ def as_dict_with_promql_labels(self):
+        self.message = message
 
-def load_alert_rule_from_file(path: Path, topology: JujuTopology) -> Optional[dict]:
-    """Load alert rule from a rules file.
+        super().__init__(self.message)
+
+
+class AlertRules:
+    """Data store for prometheus alert rules.
+
+    Alert rule files can be read in both the official and the LMA format. All rules are combined in
+    a single data structure representing an amalgamated rules file.
+
+    The official Prometheus format is a YAML file conforming to the Prometheus documentation
+    (https://prometheus.io/docs/prometheus/latest/configuration/alerting_rules/).
+    The LMA format is a subsection of the official YAML, having a single alert rule, effectively
+    "one alert per file". This class supports either format.
+
+    For the purpose of forwarding alert rules over relation data, all rules are aggregated into a
+    single data structure (which can then be written into a single *.rules YAML file).
 
     Args:
-        path: path to a *.rule file with a single rule ("groups" super section omitted).
-        topology: a `JujuTopology` instance.
+        topology: a JujuTopology instance.
     """
-    with path.open() as rule_file:
-        # Load a list of rules from file then add labels and filters
-        try:
-            rule = yaml.safe_load(rule_file)
-        except Exception as e:
-            logger.error("Failed to read alert rules from %s: %s", path.name, e)
-            return None
-        else:
-            # add "juju_" topology labels
-            if "labels" not in rule:
-                rule["labels"] = {}
-            rule["labels"].update(topology.as_dict_with_promql_labels())
 
-            if "expr" not in rule:
-                logger.error("Invalid alert rule %s: missing an 'expr' property.", path.name)
-                return None
+    # This class uses the following terminology for the various parts of a rule file:
+    # - alert rules file: the entire groups[] yaml, including the "groups:" key.
+    # - alert groups (plural): the list of groups[] (a list, i.e. no "groups:" key) - it is a list
+    #   of dictionaries that have the "name" and "rules" keys.
+    # - alert group (singular): a single dictionary that has the "name" and "rules" keys.
+    # - alert rules (plural): all the alerts in a given alert group - a list of dictionaries with
+    #   the "alert" and "expr" keys.
+    # - alert rule (singular): a single dictionary that has the "alert" and "expr" keys.
 
-            # insert juju topology filters into a prometheus alert rule
-            rule["expr"] = topology.render(rule["expr"])
-            return rule
+    def __init__(self, topology: JujuTopology):
+        self.topology = topology
+        self.alert_groups = []  # type: List[dict]
 
+    @staticmethod
+    def _is_valid_official_format(rules_dict: dict) -> bool:
+        # TODO validate against schema using e.g. pydantic
```

Comment:
Promtool already provides validation of alert rules. I am not sure if we should reimplement this functionality using pydantic. Besides schema validation for relation data is a planned feature in OF.  I do not think we should be using a TODO comment to suggest a solution that has not been discussed, agreed or spec'd out in the team or in a github ticket. Kindly create a github ticket where this can be discussed and alternatives considered. Even the TODO is probably not necessary.